### PR TITLE
DEVX-2424: Add sliding windows tutorials

### PR DESCRIPTION
**Pull request by bbejeck**
### Description
Add sliding windows tutorial a new feature in Kafka Streams 2.7

### Staging Docs

http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/DEVX-2424_sliding_windows/
http://kafka-tutorials-staging.s3-website-us-west-2.amazonaws.com/DEVX-2424_sliding_windows/sliding-windows/kstreams.html

### CODE REVIEW

**Comment by ybyzek** [comment]
@bbejeck would it be appropriate to provide a mention and link to https://cwiki.apache.org/confluence/display/KAFKA/KIP-450%3A+Sliding+Window+Aggregations+in+the+DSL from within the KT?

**Review by awalther28 — COMMENTED** [review]
@bbejeck this is awesome! I definitely learned something new reading through this. I added some comments, IMO nothing mission critical but food for thought. 
I don't know if you have any graphic on sliding windows, incorporating that could help add clarity for visual learners (like myself 😁 ).

File: _data/tutorials.yaml [hunk 1]
```diff
@@ -650,3 +650,22 @@ geo-distance:
     ruby: disabled
     scala: disabled
     swift: disabled
+
+sliding-windows:
+  title: "How to create sliding windows"
+  meta-description: "How to create sliding windows"
+  slug: "/sliding-windows"
+  question: "How do I create windowed calculations on time-series data with small advances in time?"
+  introduction: "You have time-series records and you want to create windowed aggregations with small increments in time.  You could use hopping-windows, but with small time increments, hopping windows aren't the best solution.  In this tutorial we'll cover how you can achieve efficient windowed aggregations with small advances in time"
```
[the remark this answers is not in the document]
Comment (bbejeck) [reply _data/tutorials.yaml]: I added an explanation in the `answer` section as well as the `build topology` section

File: _includes/tutorials/sliding-windows/kstreams/markup/dev/make-topology.adoc [hunk 2]
```diff
@@ -0,0 +1,34 @@
+////
+In this file you describe the Kafka streams topology, and should cover the main points of the tutorial.
+The text assumes a method buildTopology exists and constructs the Kafka Streams application.  Feel free to modify the text below to suit your needs.
+////
+
+Before you create the Kafka Streams application let's go over the key points.  In this tutorial you'll learn about using https://javadoc.io/doc/org.apache.kafka/kafka-streams/latest/org/apache/kafka/streams/kstream/SlidingWindows.html[SlidingWindows] for https://docs.confluent.io/platform/current/streams/developer-guide/dsl-api.html#aggregating[aggregation operations].  Incoming record timestamps drive the window behavior, and the time difference between records plus a grace period for out-of-order records (both are user defined) drives the size of the window.
```
[the remark this answers is not in the document]
Comment (bbejeck) [reply _includes/tutorials/sliding-windows/kstreams/markup/dev/make-topology.adoc:6]: Mentioned this the answer and build topology section when addressing https://github.com/confluentinc/kafka-tutorials/pull/690#discussion_r565617110

File: _data/harnesses/sliding-windows/kstreams.yml [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+          file: src/test/java/io/confluent/developer/SlidingWindowTest.java
+          render:
+            file: tutorials/sliding-windows/kstreams/markup/test/make-topology-test.adoc
+
+    - title: Invoke the tests
+      content:
+        - action: execute
+          file: tutorial-steps/test/invoke-tests.sh
+          render:
+            file: tutorials/sliding-windows/kstreams/markup/test/invoke-tests.adoc
```
[the remark this answers is not in the document]
Comment (bbejeck) [reply _data/harnesses/sliding-windows/kstreams.yml:175]: Good point.  Gradle makes it clear whether the tests pass or fail, but you could have a situation where the unit tests fail, but the harness runner for tutorial passes.  

I don't think we should add a step visually in the tutorial but I've added verification for the harness runner that the unit tests pass.

File: _includes/tutorials/sliding-windows/kstreams/code/Makefile [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+TEMP_DIR := $(shell mktemp -d)
+SEQUENCE := "dev, test, prod, ccloud"
+
+tutorial:
+	rm -r $(DEV_OUTPUTS_DIR) || true
+	mkdir $(DEV_OUTPUTS_DIR)
+	cp $(STEPS_DIR)/dev/datagen-temperature-reading.avsc $(TEMP_DIR)
+	harness-runner ../../../../../_data/harnesses/sliding-windows/kstreams.yml $(TEMP_DIR) $(SEQUENCE)
+	bash -c "diff --strip-trailing-cr  <(tail -n 1 $(STEPS_DIR)/dev/expected-output.txt) <(tail -n 1 $(DEV_OUTPUTS_DIR)/actual-output.txt)"
+	bash -c "grep 'BUILD SUCCESSFUL' $(DEV_OUTPUTS_DIR)/unit-test-output.txt"
```
Comment (bbejeck) [inline _includes/tutorials/sliding-windows/kstreams/code/Makefile:12]: @awalther28 added verification step here for tests cf https://github.com/confluentinc/kafka-tutorials/pull/690#discussion_r565622095

File: _includes/tutorials/sliding-windows/kstreams/markup/answer/sliding-window-answer.adoc [hunk 5]
```diff
@@ -0,0 +1,9 @@
+To use `SlidingWindows` use the `SlidingWindows.withTimeDifferenceAndGrace` method inside a `windowedBy` call.
+
++++++
+<pre class="snippet"><code class="java">{% include_raw tutorials/sliding-windows/kstreams/markup/answer/sliding-window-code.adoc %}</code></pre>
++++++
+
+The first parameter determinines the maximum time difference between records in the same window, and the second parameter sets the grace period for allowing out-of-order events into the window.  For specfics on sliding windows you can read https://cwiki.apache.org/confluence/display/KAFKA/KIP-450%3A+Sliding+Window+Aggregations+in+the+DSL[KIP-450].
+
+You can get similar behavior in hopping windows by defining a short advance time. But this will result in poor performance because hopping windows will create many overlapping, possibly redundant windows.  Performing aggregation operations over redundant windows costs CPU time, which can be expensive.  Sliding windows only create windows containing distinct items, and perform calculations on these is more efficient. Additionally, sliding windows are inclusive on both the start and end time vs. hopping windows being inclusive only on the start time.
```
Comment (bbejeck) [inline _includes/tutorials/sliding-windows/kstreams/markup/answer/sliding-window-answer.adoc:9]: @awalther28 addresses comment https://github.com/confluentinc/kafka-tutorials/pull/690#discussion_r565617110

File: _includes/tutorials/sliding-windows/kstreams/markup/dev/make-topology.adoc [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+----
+
+<1> Grouping by key, a prerequisite for aggregation operations
+<2> Specifying a sliding window for the windowing operation
+<3> The aggregation operation which sums the temperature and keeps a total count of the readings
+
+As the window "slides" over the record stream, a new window is created each time a record enters or exits the window.  Records that arrive after grace period are dropped. A window closes when `stream-time` exceeds the `window-end` + `grace-period`
+
+You can get similar behavior in hopping windows by defining a short advance time. But this will result in poor performance because hopping windows will create many overlapping, possibly redundant windows.  Performing aggregation operations over redundant windows costs CPU time, which can be expensive.  Sliding windows only create windows containing distinct items, and perform calculations on these is more efficient. Additionally, sliding windows are inclusive on both the start and end time vs. hopping windows being inclusive only on the start time.
```
Comment (bbejeck) [inline _includes/tutorials/sliding-windows/kstreams/markup/dev/make-topology.adoc:28]: @awalther28 addresses comment https://github.com/confluentinc/kafka-tutorials/pull/690#discussion_r565617110

**Comment by bbejeck** [comment]
@ybyzek  and @awalther28 updated and addressed all comments

File: _includes/tutorials/sliding-windows/kstreams/code/Makefile (hunk 4 above)
Comment (awalther28) [reply _includes/tutorials/sliding-windows/kstreams/code/Makefile:12]: Thanks @bbejeck !

File: _includes/tutorials/sliding-windows/kstreams/markup/answer/sliding-window-answer.adoc (hunk 5 above)
Comment (awalther28) [reply _includes/tutorials/sliding-windows/kstreams/markup/answer/sliding-window-answer.adoc:9]: Thanks, that makes it really clear why you wouldn't want to use hopping windows here.

**Review by awalther28 — APPROVED** [review]
Thanks for addressing my comments @bbejeck, overall LGTM!

**Pull request merged**